frame/mosaic-vault/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+
+			 T::RelayerOrigin::ensure_origin(origin)?;
+         
+			 ensure!(Self::pause_status() == false, Error::<T>::ContractPaused);
+
+			 Self::only_supported_remote_token(remote_network_id.clone(), asset_id.clone())?;
+             
+			ensure!(Self::has_been_withdrawn(deposit_id) == false, Error::<T>::AlreadyWithdrawn);
+
+			ensure!(Self::get_current_token_liquidity(asset_id)? >= amount, Error::<T>::InsufficientAssetBalance); 
+
+			  <HasBeenWithdrawn<T>>::insert(deposit_id, true);
+
+			  <LastWithdrawID<T>>::put(deposit_id);
+
+			  let pallet_account_id = Self::account_id(); 
+	
+			  let withdraw_amount = amount.saturating_sub(fee);
+
+			 T::Currency::mint_into(asset_id, &pallet_account_id, amount).map_err(|_|Error::<T>::MintToFailed)?;
+
+			 Self::decrease_total_value_transferred(asset_id, amount)?;
+
+			  T::Currency::transfer(asset_id, &pallet_account_id, &destination_account, withdraw_amount, true).map_err(|_|Error::<T>::TransferFromFailed)?;
+			
+			 if fee > T::Balance::zero() {  
+			   
+				 T::Currency::transfer(asset_id, &pallet_account_id, &Self::get_fee_address(), fee, true).map_err(|_|Error::<T>::TransferFromFailed)?;
+				
+				Self::deposit_event(Event::FeeTaken{
+					sender, 
+					destination_account: destination_account.clone(), 
+					asset_id,
+					amount,
+					fee,
+					deposit_id,
+				});
+			 }
+
+			 Self::deposit_event(Event::WithdrawalCompleted{
+				destination_account,
+				amount,
+				withdraw_amount,
+				fee,
+				asset_id,
+				deposit_id
+			 });
+
+			 Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn unlock_in_transfer_funds(
+			origin: OriginFor<T>,
+			asset_id: T:: AssetId,
+			amount: T::Balance,
+			deposit_id: T::DepositId,
+		 ) ->DispatchResultWithPostInfo {
+
+			T::RelayerOrigin::ensure_origin(origin)?;
+
+			ensure!(Self::pause_status() == false, Error::<T>::ContractPaused);
+			
+			ensure!(Self::has_been_completed(deposit_id) == false, Error::<T>::AlreadCompleted);
+
+			ensure!(Self::in_transfer_funds(asset_id) >= amount, Error::<T>::InsufficientFunds);
+
+			let deposit = Self::deposits(asset_id);
+
+			ensure!(deposit.asset_id == asset_id && deposit.amount == amount, Error::<T>::InsufficientFunds);
+
+			<HasBeenCompleted<T>>::insert(deposit_id, true);
+
+	       let new_intransfer_funds = Self::in_transfer_funds(asset_id).checked_sub(&amount).ok_or(Error::<T>::Underflow)?;
+
+		   <InTransferFunds<T>>::insert(asset_id, new_intransfer_funds);
+
+		   Self::deposit_event(Event::TransferFundsUnlocked{asset_id, amount, deposit_id});
+	
+			Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn unlock_funds(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			user_account_id: T::AccountId,
+			amount: T::Balance,
+			deposit_id: T::DepositId,
+		 ) ->DispatchResultWithPostInfo {
+            
+			 T::RelayerOrigin::ensure_origin(origin.clone())?;
+          
+			 ensure!(Self::has_been_unlocked(deposit_id) == false, Error::<T>::AssetUnlreadyUnlocked);
```

Comment:
you should not check for being unlocked, but remove unlocked from state i think. yep, it will be not clear from this call if funds ever been locked or unlocked. but that is performance.

Reply:
The has_been_unlocked state is used to stop execution for the same id, so if removed, how is the check handled